### PR TITLE
Klee CMake instructions

### PR DESCRIPTION
**Pull request by delcypher**
I don't have any desire to confuse people by giving them an option when using LLVM 3.4 of both build systems. The new build system works in more scenarios (e.g. you can build against an installed LLVM on non Debian distributions) than the old one so this should be preferred as it is more likely to work.

### CODE REVIEW

**Comment by andreamattavelli** [comment]
@delcypher Besides the discussion of what build system to use as default, I vote for leaving a copy of the build instructions with Autoconf/Makefile. In fact, we should provide help to whoever wants to (or is even _forced_ to) use the Autoconf/Makefile build system.

**Comment by delcypher** [comment]
@andreamattavelli The build instructions for the Autoconf/Makefile are a little out of date and I really don't want to maintain yet another set of instructions.

I'd rather just make this change now because I want new users to use the new build system. If people are using the old build system that's fine but if they are using it now then they already know how to use it and don't need this documentation.

**Comment by delcypher** [comment]
> (or is even forced to) use the Autoconf/Makefile build system.

What scenario is that? This is documentation is for KLEE in the master branch. It is **not** for people who have forked KLEE and modified it. We are not beholden to downstream users.

**Comment by delcypher** [comment]
Anyway I'll let @ccadar @MartinNowack have their say and go with what the majority want.

**Comment by andreamattavelli** [comment]
@delcypher I don't see why we shouldn't, since I am not aware of any concrete plan on removing Autoconf/Makefile.

**Comment by delcypher** [comment]
@andreamattavelli Good point about there not being a concrete plan. I'll create an issue about this.

**Comment by MartinNowack** [comment]
@delcypher I second @andreamattavelli.
Keep the old documentation, if people want to build the old version (e.g. for comparison or to reproduce results from published papers, they will need this documentation).
Also LLVM keeps their documentation for every different versions. This is quite often still useful for me.

Beside that, up to 3.8, those build instructions work well.
If they don't, people will complain and we can fix it.

**Comment by ccadar** [comment]
I've just seen this -- my recent comment in https://github.com/klee/klee/issues/496 seems in line with what the majority wants.

**Comment by andreamattavelli** [comment]
> Also LLVM keeps their documentation for every different versions. This is quite often still useful for me.

Right, that's what I would also do here.

@delcypher To clarify my point, I'm not arguing against CMake at all. On the contrary, I see CMake as KLEE's future build system. I am just saying that the website is the KLEE's documentation, and as such I would leave there everything. In fact, I will express the same ideas when we will drop support for LLVM 2.9.

**Comment by delcypher** [comment]
Okay I've redone the docs and I've kept the old documentation for LLVM 3.4 with the old build system.

I've also dropped saying LLVM 3.4 is experimental. That doesn't make sense given that the web page says that using LLVM 3.4 is recommended. I also use LLVM 3.4 heavily so it is being supported.

File: build-llvm34.md [hunk 1]
```diff
@@ -50,13 +41,15 @@ If you want to build KLEE with LLVM 2.9, [click here]({{site.baseurl}}/build-llv
...
 
-   ```bash
-   $ sudo ln -sf /usr/bin/llvm-config-3.4 /usr/bin/llvm-config
-   ```
-
    That's it for LLVM. If you want to install it manually, please refer to the official [LLVM Getting Started documentation](http://www.llvm.org/docs/GettingStarted.html).<br/><br/>  
 
-3. **Build STP:** KLEE is based on the STP constraint solver, you can find the instructions [here]({{site.baseurl}}/build-stp).
+3. **Install constraint solver(s)**
+
```
Comment (ccadar) [inline build-llvm34.md]: We should still link to build-stp.md

File: build-llvm34.md [hunk 2]
```diff
@@ -6,8 +6,12 @@ slug: build-llvm34
 ---
 
 The current procedure for building KLEE with LLVM 3.4 (recommended) is outlined below.
+These instructions use KLEE's new CMake based build system. If you wish to build with
+KLEE's older build system (not recommeneded) then see [here]({{site.baseurl}}/build-llvm34-legacy).
```
Comment (delcypher) [inline build-llvm34.md]: Oops. Thanks for spotting that.

File: build-llvm34.md [hunk 3]
```diff
@@ -68,79 +61,126 @@ If you want to build KLEE with LLVM 2.9, [click here]({{site.baseurl}}/build-llv
...
+   $ make  
+   ```
+
+9. **(Optional) Run the main regression test suite**
+   If you installed `lit` and KLEE was configured with integration tests enabled
+   then you can run them like this.
    
    ```bash
-   $ make test
+   $ make integrationtests
```
Comment (andreamattavelli) [inline build-llvm34.md]: We always used `make test` in the other build instructions.

**Comment by andreamattavelli** [comment]
Looks good to me, besides that couple of minor comments.
Unfortunately the end result is less readable than expected: sometimes there is no space between items, but it is a theme-related issue..

File: build-llvm34.md (hunk 3 above)
Comment (delcypher) [inline build-llvm34.md]: The CMake build system uses a different target name. This was a deliberate choice because `test` is a very vague target name. The old build system had the unit test target called `unittests` so I preserved that name for the unit tests and used a similar name for the integration tests.

Comment (andreamattavelli) [inline build-llvm34.md]: I see. However, those are not integration tests but system tests In my opinion. In fact, you don't tests units together but the system as a whole. Maybe `regressiontests` is more appropriate, although still not completely correct.

Comment (andreamattavelli) [inline build-llvm34.md]: You're right.

**Comment by andreamattavelli** [comment]
@delcypher Since we are already modifying the website, in `docs/index.md` I would modify these lines
```markdown
1. [Building KLEE (LLVM 2.9)]({{site.baseurl}}/build-llvm29/): Instructions on how to build KLEE from source using LLVM 2.9.                                                                                                       
1. [Building KLEE (LLVM 3.4)]({{site.baseurl}}/build-llvm34/): Instructions on how to build KLEE from source using LLVM 3.4.   
```

into something like this:

```markdown
1. [Building KLEE (LLVM 3.4) with CMake]({{site.baseurl}}/XXX/): Instructions on how to build KLEE from source using LLVM 3.4 and CMake.
1. [Building KLEE (LLVM 3.4) with Makefiles]({{site.baseurl}}/YYY/): Instructions on how to build KLEE from source using LLVM 3.4.
1. [Building KLEE (LLVM 2.9)]({{site.baseurl}}/build-llvm29/): Instructions on how to build KLEE from source using LLVM 2.9.
```

File: build-llvm34.md (hunk 3 above)
Comment (delcypher) [inline build-llvm34.md]: @andreamattavelli 
>  as a user we are calling the same identical operation in two ways.
Now you've confused me. That is not correct.

There are three testing targets.

* `integrationtests` - Run integration tests
* `unittests` - Build and run unittests
* `check` - Run both of the above. This is convenience target to run all types of test. We only have two types now but we could add more in the future.

They all do different things.

File: docs/index.md [hunk 4]
```diff
@@ -6,8 +6,9 @@ slug: documentation
 ---
 
 1. [Using KLEE Docker image]({{site.baseurl}}/docker/): Instructions on how to use the KLEE Docker image.
+1. [Building KLEE (LLVM 3.4)]({{site.baseurl}}/build-llvm34/): Instructions on how to build KLEE from source using LLVM 3.4 and CMake.
+1. [Building KLEE (LLVM 3.4) using legacy build system]({{site.baseurl}}/build-llvm34-legacy/): Instructions on how to build KLEE from source using LLVM 3.4 and KLEE's legacy build system.
```
Comment (ccadar) [inline docs/index.md]: I would call this autoconf, not legacy build system

File: getting-started.md [hunk 5]
```diff
@@ -9,4 +9,5 @@ There are three ways to get started with KLEE.
 
 * [Use our Docker images]({{site.baseurl}}/docker): this is the fastest way to get started.
 * [Build from source against LLVM 3.4]({{site.baseurl}}/build-llvm34): this is the current recommended version.
+* [Build from source against LLVM 3.4 (legacy build system)]({{site.baseurl}}/build-llvm34): Same as above but using legacy build system.
```
Comment (ccadar) [inline getting-started.md]: Same here

File: build-llvm34.md [hunk 6]
```diff
@@ -6,32 +6,23 @@ slug: build-llvm34
 ---
 
 The current procedure for building KLEE with LLVM 3.4 (recommended) is outlined below.
+These instructions use KLEE's new CMake based build system. If you wish to build with
+KLEE's older build system (not recommended) then see [here]({{site.baseurl}}/build-llvm34-legacy).
```
Comment (ccadar) [inline build-llvm34.md]: Say autoconf here, drop "not recommended", as it's a bit strong, we're still maintaining the autoconf system.

File: build-llvm34.md [hunk 7]
```diff
@@ -50,13 +41,15 @@ If you want to build KLEE with LLVM 2.9, [click here]({{site.baseurl}}/build-llv
...
-   ```
-
    That's it for LLVM. If you want to install it manually, please refer to the official [LLVM Getting Started documentation](http://www.llvm.org/docs/GettingStarted.html).<br/><br/>  
 
-3. **Build STP:** KLEE is based on the STP constraint solver, you can find the instructions [here]({{site.baseurl}}/build-stp).
+3. **Install constraint solver(s)**
+
+   KLEE supports multiple different constraint solvers. You must install at least one to build KLEE.
+
+   * [STP](https://github.com/stp/stp) Historically KLEE was built around only this solver so support for this solver is the most stable. For build instructions see [here]({{site.baseurl}}/build-stp). When configuring KLEE in step 9 pass `-DENABLE_SOLVER_STP=ON` to CMake to enable support for STP. If you don't have STP installed but have a built copy you can also pass `-DSTP_DIR=<STP_BUILD_DIR>` to CMake when configuring KLEE where `<STP_BUILD_DIR>` is the absolute path to the STP build directory which contains the file `STPConfig.cmake`.
```
Comment (ccadar) [inline build-llvm34.md]: I think we should move all these references to Step 9 to Step 9 itself.  I found it hard to remember all this in Step 3.

**Comment by ccadar** [comment]
I find it quite hard to follow these instructions compared to the previous ones.  I think we should just show users one standard way to install KLEE (e.g., just use STP), rather than presenting all possible options on that page.

After trying to follow them, I get this error while running make.  What did I miss?  

```
[ 95%] Building CXX object tools/klee/CMakeFiles/klee.dir/main.cpp.o
/data/klee/klee-3.4/tools/klee/main.cpp: In static member function ‘static std::string KleeHandler::getRunTimeLibraryPath(const char*)’:
/data/klee/klee-3.4/tools/klee/main.cpp:610:15: error: ‘KLEE_INSTALL_BIN_DIR’ was not declared in this scope
   if (strlen( KLEE_INSTALL_BIN_DIR ) != 0 &&
```

**Comment by delcypher** [comment]
> After trying to follow them, I get this error while running make. What did I miss?

```
[ 95%] Building CXX object tools/klee/CMakeFiles/klee.dir/main.cpp.o
/data/klee/klee-3.4/tools/klee/main.cpp: In static member function ‘static std::string KleeHandler::getRunTimeLibraryPath(const char*)’:
/data/klee/klee-3.4/tools/klee/main.cpp:610:15: error: ‘KLEE_INSTALL_BIN_DIR’ was not declared in this scope
   if (strlen( KLEE_INSTALL_BIN_DIR ) != 0 &&
```

Check your generated `include/klee/Config/config.h` in the build directory. The macro is defined there.

* Are you sure you're using KLEE's latest sources on the master branch?
* Make sure that `include/klee/Config/config.h` does not exist in your source tree. If you did an in source build with the old build system there that will actually get picked up in preference to the one in the build directory. If that is a case let me know because that's a bug that I should fix.

File: build-llvm34.md (hunk 7 above)
Comment (delcypher) [inline build-llvm34.md]: I see what you mean. It was the easiest way to write it but it doesn't make very easy reading. I'll try to rewrite it.

**Comment by delcypher** [comment]
> Make sure that include/klee/Config/config.h does not exist in your source tree. If you did an in source build with the old build system there that will actually get picked up in preference to the one in the build directory. If that is a case let me know because that's a bug that I should fix.

Turns out that was a bug. I've made a fix in https://github.com/klee/klee/pull/501

**Comment by delcypher** [comment]
I've reworked the instructions.

File: build-llvm34.md (hunk 3 above)
Comment (andreamattavelli) [inline build-llvm34.md]: Nope: there are check, test, and unittests. That's the current state of KLEE and I would push for consistency at the moment. If we want to change name, then we need to change name in both build systems.

**Comment by andreamattavelli** [comment]
@delcypher The instructions are clear in my opinion. I leave @ccadar to review his comments.

**Comment by ccadar** [comment]
I think it's almost there.  I'd still like to do a full pass over them (btw, is there an embedded .md viewer in GitHub?) but more importantly I'd like to first settle the test targets issue before going live with them.  I guess the right place to discuss this is in https://github.com/klee/klee/issues/500, after which we can come back here.

**Comment by delcypher** [comment]
@ccadar I don't think there is an embedded viewer. It's better to just grab the sources and run jekyll locally on your machine to see the final output.

File: build-llvm34.md [hunk 8]
```diff
@@ -70,77 +63,130 @@ If you want to build KLEE with LLVM 2.9, [click here]({{site.baseurl}}/build-llv
...
 
    ```bash
-   $ make  
+   cmake \
+     -DENABLE_SOLVER_STP=ON \
+     -DENABLE_POSIX_RUNTIME=ON \
+     -DENABLE_KLEE_UCLIBC \
+     -DKLEE_UCLIBC_PATH=<KLEE_UCLIBC_SOURCE_DIR> \
+     -DGTEST_SRC_DIR=<GTEST_SOURCE_DIR> \
+     -DENABLE_TESTS=ON \
```
Comment (ccadar) [inline build-llvm34.md]: I find these options confusing.  Why do we have a global `-DENABLE_TESTS` and also `-DENABLE_UNIT_TESTS` and `-DENABLE_UNIT_TESTS`?   Also, if the defaults are `ON`, we should show them here.

Comment (delcypher) [inline build-llvm34.md]: I'm confused.

This is the cmake invocation:

```
cmake \
  -DENABLE_SOLVER_STP=ON \
  -DENABLE_POSIX_RUNTIME=ON \
  -DENABLE_KLEE_UCLIBC \
  -DKLEE_UCLIBC_PATH=<KLEE_UCLIBC_SOURCE_DIR> \
  -DGTEST_SRC_DIR=<GTEST_SOURCE_DIR> \
  -DENABLE_TESTS=ON \
  -DENABLE_INTEGRATION_TESTS=ON \
  -DENABLE_UNIT_TESTS=ON \
  <KLEE_SRC_DIRECTORY>
```

>  Why do we have a global -DENABLE_TESTS and also -DENABLE_UNIT_TESTS and -DENABLE_UNIT_TESTS?

`-DENABLE_UNIT_TESTS` is only listed once. There are currently three testing options. `ENABLE_TESTS` allows all tests to be globally enabled/disabled. The other options (`ENABLE_INTEGRATION_TESTS` and `ENABLE_UNIT_TESTS`) allow different types of test to be enabled/disabled. These are all documented in `README-CMake.md`.

I see nothing confusing about that.

> Also, if the defaults are ON, we should show them here.
The defaults **are ON** and I'm showing them in the invocation to be explicit.

Comment (ccadar) [inline build-llvm34.md]: Just to clarify my confusion with the `TEST` option.  My point is that we don't need a global `-DENABLE_UNIT_TESTS` if we have the more fined-grained `-DENABLE_UNIT_TESTS` and `-DENABLE_UNIT_TESTS`.  Or let me ask this: what happens if I set `-DENABLE_UNIT_TESTS=OFF` and the other two to `ON`?

File: build-llvm34-legacy.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
+---
```
Comment (ccadar) [inline build-llvm34-legacy.md]: Can you please rename this file to `build-llvm34-autoconf.md`?

File: build-llvm34.md (hunk 8 above)
Comment (delcypher) [inline build-llvm34.md]: I'm still really confused.

>  My point is that we don't need a global -DENABLE_UNIT_TESTS if we have the more fined-grained -DENABLE_UNIT_TESTS and -DENABLE_UNIT_TESTS

That makes no sense.

File: build-llvm34.md (hunk 3 above)
Comment (delcypher) [inline build-llvm34.md]: I can add an alias to both systems... I think. I'll have to try it.

Disabling unit tests is only an option in CMake but I'll go test this now.

File: build-llvm34.md (hunk 8 above)
Comment (ccadar) [inline build-llvm34.md]: Sorry, copy and paste error, can you re-read my message now?

File: build-llvm34.md (hunk 3 above)
Comment (delcypher) [inline build-llvm34.md]: - [X] (1) Ensure `make check` works if unit tests are disabled.
- [ ] (2) Have make test be an alias for make check. I just tried and this is **not going to work**. `test` is a reserved target name in CMake for its own test infrastructure (`CTest`) which I don't want to use. So we can't do this.
- [X] (3) Have the same targets in both build systems. The targets `unittests`, `check` and `systemtests` now run the same tests in both build systems.

Comment (ccadar) [inline build-llvm34.md]: OK, I'll live without `make test` if it's a big no-no in CMake

File: build-llvm34-legacy.md (hunk 9 above)
Comment (delcypher) [inline build-llvm34-legacy.md]: Done

**Comment by delcypher** [comment]
Instructions for both build systems are now in sync.

**Comment by ccadar** [comment]
Fantastic! If you also copy the current build instructions into a page for KLEE 1.3, I'm happy to merge everything :)

File: build-llvm34.md (hunk 8 above)
Comment (ccadar) [inline build-llvm34.md]: (Typo fixed).  OK, glad we're on the same page here.  OK, let's change it later, as I really think they're complicating things.
